iotdb/src/main/java/org/apache/iotdb/db/query/dataset/EngineDataSetWithTimeGenerator.java

```diff
@@ -52,24 +53,49 @@ public EngineDataSetWithTimeGenerator(List<Path> paths, List<TSDataType> dataTyp
 
   @Override
   public boolean hasNext() throws IOException {
-    return timeGenerator.hasNext();
+    if (cachedRowRecord != null) {
+      return true;
+    }
+    return cacheRowRecord();
   }
 
   @Override
   public RowRecord next() throws IOException {
-    long timestamp = timeGenerator.next();
-    RowRecord rowRecord = new RowRecord(timestamp);
-    for (int i = 0; i < readers.size(); i++) {
-      EngineReaderByTimeStamp reader = readers.get(i);
-      Object value = reader.getValueInTimestamp(timestamp);
-      if (value == null) {
-        rowRecord.addField(new Field(null));
-      } else {
-        rowRecord.addField(getField(value, dataTypes.get(i)));
-      }
+    if (cachedRowRecord == null) {
+      cacheRowRecord();
     }
+    RowRecord tempRecord = cachedRowRecord;
+    cachedRowRecord = null;
+    return tempRecord;
+  }
 
-    return rowRecord;
+  /**
+   * Cache row record
+   *
+   * @return if there has next row record.
+   */
+  private boolean cacheRowRecord() throws IOException {
+    while (timeGenerator.hasNext()) {
+      boolean markNull = true;
```

Comment:
Suggest to use another val name such as `hasNonNullField` with a initial value `false`

Reply:
accept this advice